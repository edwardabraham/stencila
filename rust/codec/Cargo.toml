[package]
name = "codec"
description = "Defines a specification and trait for codecs"
version = "0.0.0"
edition = "2021"

[dependencies]
common = { path = "../common" }
stencila-schema = { path = "../schema" }
utils = { path = "../utils" }
