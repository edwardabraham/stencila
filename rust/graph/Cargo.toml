[package]
name = "graph"
version = "0.0.0"
edition = "2021"

[dependencies]
derivative = "2.2.0"
eyre = "0.6.5"
graph-triples = { path = "../graph-triples", version = "0.0.0" }
path-slash = "0.1.4"
petgraph = { version = "0.6.0", features = ["serde-1"] }
schemars = { version = "0.8.6", features = ["preserve_order"] }
serde = "1.0.130"
serde_json = { version = "1.0.68", features = ["preserve_order"] }
serde_with = "1.11.0"
strum = { version = "0.22.0", features = ["derive"] }
utils = { path = "../utils", version = "0.0.0" }
