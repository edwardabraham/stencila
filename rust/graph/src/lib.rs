use eyre::{bail, Result};
use graph_triples::{direction, relations, Direction, Pairs, Relation, Resource, Triple};
use path_slash::PathExt;
use petgraph::{
    graph::NodeIndex,
    stable_graph::StableGraph,
    visit::{EdgeRef, IntoEdgeReferences, IntoNodeReferences},
};
use schemars::{
    gen::SchemaGenerator,
    schema::{Schema, SchemaObject},
    schema_for, JsonSchema,
};
use serde::{ser::SerializeMap, Serialize};
use serde_json::json;
use serde_with::skip_serializing_none;
use std::{
    collections::HashMap,
    path::{Path, PathBuf},
};
use strum::Display;
use utils::some_string;

/// A dependency graph for a project or document
#[derive(Debug, Default, Clone)]
pub struct Graph {
    /// The path of the project or document that this graph is for
    ///
    /// Primarily used to make file paths relative in visualizations and
    /// if ever persisting the graph.
    path: PathBuf,

    /// The graph itself
    ///
    /// Use a `petgraph::StableGraph` so that nodes can be added and removed
    /// without changing node indices.
    graph: StableGraph<Resource, Relation>,

    /// Indices of the nodes in the tree
    ///
    /// This is necessary to keep track of which resources
    /// are already in the graph and re-use their index if they are.
    indices: HashMap<Resource, NodeIndex>,
}

impl Serialize for Graph {
    /// Custom serialization to strip prefix from paths, add stable node indices,
    /// and exclude properties that are included by default by `petgraph` (e.g `node_holes`).
    ///
    /// Our general approach is to keep paths absolute whilst in memory and only convert to
    /// relative when necessary (e.g. visualizations). See also `Graph::to_dot`.
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let nodes: Vec<serde_json::Value> = self
            .graph
            .node_references()
            .map(|(index, resource)| {
                let mut obj = serde_json::to_value(resource).expect("To be able to serialize");
                let obj = obj.as_object_mut().expect("To be an object");

                // Strip prefix from paths
                if let Some(path) = match resource {
                    Resource::Symbol(symbol) => Some(symbol.path.clone()),
                    Resource::Node(node) => Some(node.path.clone()),
                    Resource::File(file) => Some(file.path.clone()),
                    _ => None,
                } {
                    let path = path
                        .strip_prefix(&self.path)
                        .unwrap_or(&path)
                        .to_slash_lossy();
                    obj.insert("path".to_string(), json!(path));
                }

                obj.insert("index".to_string(), json!(index));
                json!(obj)
            })
            .collect();

        let edges: Vec<serde_json::Value> = self
            .graph
            .edge_references()
            .map(|edge| -> serde_json::Value {
                json!({
                    "from": edge.source(),
                    "to": edge.target(),
                    "relation": edge.weight()
                })
            })
            .collect();

        let mut map = serializer.serialize_map(Some(2))?;
        map.serialize_entry("nodes", &nodes)?;
        map.serialize_entry("edges", &edges)?;
        map.end()
    }
}

/// The available graph serialization formats
pub const FORMATS: [&str; 3] = ["dot", "json", "yaml"];

impl Graph {
    /// Create a new, empty graph
    pub fn new<P: AsRef<Path>>(path: P) -> Graph {
        Graph {
            path: PathBuf::from(path.as_ref()),
            indices: HashMap::new(),
            graph: StableGraph::new(),
        }
    }

    /// Create a graph from set of dependency relations
    pub fn from_relations<P: AsRef<Path>>(path: P, relations: &[(Resource, Pairs)]) -> Graph {
        let mut graph = Graph::new(path);
        for (subject, pairs) in relations {
            for (relation, object) in pairs {
                graph.add_triple((subject.clone(), relation.clone(), object.clone()));
            }
        }
        graph
    }

    /// Add a resource to the graph
    pub fn add_resource(&mut self, resource: Resource) {
        if self.indices.get(&resource).is_none() {
            let index = self.graph.add_node(resource.clone());
            self.indices.insert(resource, index);
        }
    }

    /// Add a triple to the graph
    pub fn add_triple(&mut self, (subject, relation, object): Triple) {
        let subject = if let Some(index) = self.indices.get(&subject) {
            *index
        } else {
            let index = self.graph.add_node(subject.clone());
            self.indices.insert(subject, index);
            index
        };

        let object = if let Some(index) = self.indices.get(&object) {
            *index
        } else {
            let index = self.graph.add_node(object.clone());
            self.indices.insert(object, index);
            index
        };

        let (from, to) = match direction(&relation) {
            Direction::From => (object, subject),
            Direction::To => (subject, object),
        };

        self.graph.add_edge(from, to, relation);
    }

    /// Add a set of triples to the graph
    pub fn add_triples(&mut self, triples: Vec<Triple>) {
        triples
            .into_iter()
            .for_each(|triple| self.add_triple(triple))
    }

    /// Convert the graph to some format
    pub fn to_format(&self, format: &str) -> Result<String> {
        Ok(match format {
            "dot" => self.to_dot(),
            "json" => serde_json::to_string_pretty(self)?,
            "yaml" => serde_yaml::to_string(self)?,
            _ => bail!("Unknown graph format '{}'", format),
        })
    }

    /// Convert the graph to a visualization nodes and edges
    pub fn to_dot(&self) -> String {
        let nodes = self
            .indices
            .iter()
            .map(|(resource, node)| {
                let index = node.index();

                let path = match resource {
                    Resource::Symbol(symbol) => symbol.path.clone(),
                    Resource::Node(node) => node.path.clone(),
                    Resource::File(file) => file.path.clone(),
                    _ => PathBuf::new(),
                };
                let path = path
                    .strip_prefix(&self.path)
                    .unwrap_or(&path)
                    .to_slash_lossy();

                let (shape, fill_color, label) = match resource {
                    Resource::Symbol(symbol) => (
                        "diamond",
                        "#efb8b8",
                        format!(
                            "{}{}",
                            if !symbol.kind.is_empty() {
                                format!("{}\\n", symbol.kind)
                            } else {
                                "".to_string()
                            },
                            symbol.name
                        ),
                    ),
                    Resource::Node(node) => {
                        let label = if !node.id.starts_with('_') {
                            format!("{}\\n{}", node.kind, node.id)
                        } else {
                            node.kind.clone()
                        };
                        ("box", "#efe0b8", label)
                    }
                    Resource::File(..) => ("note", "#d1efb8", path.clone()),
                    Resource::Source(source) => ("house", "#efb8d4", source.name.clone()),
                    Resource::Module(module) => ("invhouse", "#b8efed", module.name.clone()),
                    Resource::Url(url) => ("box", "#cab8ef", url.url.clone()),
                };

                let node = match resource {
                    Resource::File(..) => format!(
                        r#"  n{index} [shape="point", style="invis" label="{label}"]"#,
                        index = index,
                        label = label
                    ),
                    _ => format!(
                        r#"  n{index} [shape="{shape}" fillcolor="{fill_color}" label="{label}"]"#,
                        index = index,
                        shape = shape,
                        fill_color = fill_color,
                        label = label.replace('\"', "\\\"")
                    ),
                };

                (path, node)
            })
            .collect::<Vec<(String, String)>>();

        let mut clusters: HashMap<String, Vec<String>> = HashMap::new();
        for (path, node) in nodes {
            clusters.entry(path).or_default().push(node)
        }

        let path_to_cluster = |path: &Path| {
            clusters
                .keys()
                .position(|key| {
                    key == &path
                        .strip_prefix(&self.path)
                        .unwrap_or(path)
                        .to_slash_lossy()
                })
                .unwrap_or(0)
        };

        let subgraphs = clusters
            .iter()
            .enumerate()
            .map(|(index, (label, nodes))| {
                if label.is_empty() {
                    nodes.join("\n")
                } else {
                    [
                        &format!("  subgraph cluster{} {{\n", index),
                        &format!("    label=\"{}\" fillcolor=\"#d1efb8\"\n  ", label),
                        &nodes.join("\n  "),
                        "\n  }",
                    ]
                    .concat()
                }
            })
            .collect::<Vec<String>>()
            .join("\n");

        let edges = self
            .graph
            .edge_indices()
            .filter_map(|edge| {
                if let (Some((from, to)), Some(relation)) = (
                    self.graph.edge_endpoints(edge),
                    self.graph.edge_weight(edge),
                ) {
                    let (label, style) = match relation {
                        Relation::Convert(relations::Convert { auto: active })
                        | Relation::Import(relations::Import { auto: active }) => (
                            relation.to_string(),
                            if *active { "solid" } else { "dashed" },
                        ),
                        Relation::Assign(relations::Assign { range })
                        | Relation::Use(relations::Use { range })
                        | Relation::Read(relations::Read { range })
                        | Relation::Write(relations::Write { range }) => {
                            let label = if *range == relations::NULL_RANGE {
                                relation.to_string()
                            } else {
                                format!("{} (L{})", relation, range.0 + 1)
                            };
                            (label, "solid")
                        }
                        _ => (relation.to_string(), "solid"),
                    };

                    let ltail = if let Some(Resource::File(file)) = self.graph.node_weight(from) {
                        format!(" ltail=\"cluster{}\"", path_to_cluster(&file.path))
                    } else {
                        "".to_string()
                    };

                    let lhead = if let Some(Resource::File(file)) = self.graph.node_weight(to) {
                        format!(" lhead=\"cluster{}\"", path_to_cluster(&file.path))
                    } else {
                        "".to_string()
                    };

                    Some(format!(
                        r#"  n{from} -> n{to} [label="{label}" style="{style}"{ltail}{lhead}]"#,
                        from = from.index(),
                        to = to.index(),
                        label = label,
                        style = style,
                        ltail = ltail,
                        lhead = lhead,
                    ))
                } else {
                    None
                }
            })
            .collect::<Vec<String>>()
            .join("\n");

        format!(
            r#"digraph {{
  graph [rankdir=LR compound=true fontname=Helvetica fontsize=12 labeljust=l color=gray]
  node [style=filled fontname=Helvetica fontsize=11]
  edge [fontname=Helvetica fontsize=10]

{subgraphs}

{edges}
}}
"#,
            subgraphs = subgraphs,
            edges = edges
        )
    }
}

#[derive(Debug, Display, JsonSchema, Serialize)]
#[serde(rename_all = "lowercase")]
#[strum(serialize_all = "lowercase")]
pub enum GraphEventType {
    Updated,
}

#[skip_serializing_none]
#[derive(Debug, JsonSchema, Serialize)]
#[schemars(deny_unknown_fields)]
pub struct GraphEvent {
    /// The path of the project (absolute)
    project: PathBuf,

    /// The type of event
    #[serde(rename = "type")]
    type_: GraphEventType,

    /// The graph at the time of the event
    #[schemars(schema_with = "GraphEvent::graph_schema")]
    graph: Graph,
}

impl GraphEvent {
    /// Generate the JSON Schema for the `graph` property
    fn graph_schema(_generator: &mut SchemaGenerator) -> Schema {
        Schema::Object(SchemaObject {
            reference: some_string!("Graph"),
            ..Default::default()
        })
    }

    /// Publish a `GraphEvent`.
    ///
    /// Will publish an event under the `projects:<project>:graph` topic.
    pub fn publish(project: &Path, type_: GraphEventType, graph: &Graph) {
        let topic = &format!("projects:{}:graph", project.display());
        let event = GraphEvent {
            project: project.to_path_buf(),
            type_,
            graph: graph.clone(),
        };
        events::publish(topic, &event)
    }
}

/// Get JSON Schemas for this crate
pub fn schemas() -> Result<serde_json::Value> {
    Ok(json!([
        schema_for!(Resource),
        schema_for!(Relation),
        schema_for!(GraphEvent),
        serde_json::json!({
            "$id": "Triple",
            "title": "Triple",
            "description": "A subject-relation-object triple",
            "type" : "array",
            "items": [
                {
                    "tsType": "Resource"
                },
                {
                    "tsType": "Relation"
                },
                {
                    "tsType": "Resource"
                }
            ],
            "minItems": 3,
            "maxItems": 3
        }),
        serde_json::json!({
            "$id": "Graph",
            "title": "Graph",
            "description": "A project dependency graph",
            "type" : "object",
            "required": ["nodes", "edges"],
            "properties": {
                "nodes": {
                    "description": "The resources in the graph",
                    "type": "array",
                    "items": {
                        "tsType": "Resource"
                    },
                    "isRequired": true
                },
                "edges": {
                    "description": "The relations between resources in the graph",
                    "type": "array",
                    "items": {
                        "type": "object",
                        "required": ["from", "to", "relation"],
                        "properties": {
                            "from": "integer",
                            "to": "integer",
                            "relation" : {
                                "tsType": "Resource"
                            }
                        },
                        "additionalProperties": false
                    },
                    "isRequired": true
                }
            },
            "additionalProperties": false
        }),
    ]))
}
