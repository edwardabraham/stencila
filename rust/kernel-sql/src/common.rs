use kernel::{
    common::itertools::Itertools,
    stencila_schema::{Datatable, ValidatorTypes},
};

/// Generate a `CREATE TABLE` statement for a Stencila [`Datatable`]
pub(crate) fn create_table_from_datatable(
    name: &str,
    datatable: &Datatable,
    or_replace: bool,
) -> String {
    let columns = datatable
        .columns
        .iter()
        .map(|column| {
            let item_validator = column
                .validator
                .as_deref()
                .and_then(|array_validator| array_validator.items_validator.clone());
            let data_type = match item_validator.as_deref() {
                Some(ValidatorTypes::BooleanValidator(..)) => "BOOLEAN",
                Some(ValidatorTypes::IntegerValidator(..)) => "INTEGER",
                Some(ValidatorTypes::NumberValidator(..)) => "REAL",
                Some(ValidatorTypes::StringValidator(..)) => "TEXT",
                Some(ValidatorTypes::DateValidator(..)) => "DATE",
                Some(ValidatorTypes::TimeValidator(..)) => "TIME",
                Some(ValidatorTypes::DateTimeValidator(..)) => "TIMESTAMP",
                Some(ValidatorTypes::TimestampValidator(..)) => "TIMESTAMP",
                Some(ValidatorTypes::DurationValidator(..)) => "INTERVAL",
                _ => "JSON",
            };

            let is_nullable = if let Some(validator) = &column.validator {
                if !validator.items_nullable {
                    "NOT NULL".to_string()
                } else {
                    String::new()
                }
            } else {
                String::new()
            };

            format!("{} {} {}", column.name, data_type, is_nullable)
        })
        .collect_vec()
        .join(", ");

    format!(
        "CREATE {} TABLE \"{}\"({});\n",
        if or_replace { "OR REPLACE" } else { "" },
        name,
        columns,
    )
}

/// Get the number of rows and columns of a Stencila [`Datatable`]
pub(crate) fn datatable_rows_cols(datatable: &Datatable) -> (usize, usize) {
    let rows = datatable
        .columns
        .first()
        .map(|column| column.values.len())
        .unwrap_or(0);

    let cols = datatable.columns.len();

    (rows, cols)
}
