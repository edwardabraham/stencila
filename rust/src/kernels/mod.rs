use eyre::Result;
use schemars::JsonSchema;
use serde::Serialize;
use stencila_schema::Node;

trait KernelTrait {
    fn get(&mut self, name: &str) -> Result<Node>;

    fn set(&mut self, name: &str, value: Node) -> Result<()>;

    fn exec(&mut self, code: &str) -> Result<Vec<Node>>;
}

mod calc;
mod default;

#[derive(Debug, Clone, JsonSchema, Serialize)]
pub enum Kernel {
    Default(default::DefaultKernel),
    Calc(calc::CalcKernel),
}
