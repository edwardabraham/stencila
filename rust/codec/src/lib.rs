use std::{fs::create_dir_all, path::Path};

use common::{
    async_trait::async_trait,
    eyre::{bail, Result},
    serde::{Deserialize, Serialize},
    tokio::{
        fs::File,
        io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt, BufReader, BufWriter},
    },
};
use stencila_schema::Node;
use utils::vec_string;

// Re-export for the convenience of crates that implement `CodecTrait`
pub use common;
pub use stencila_schema;
pub use utils;

/// A specification for codecs
///
/// All codecs, including those implemented in plugins, should provide this
/// specification. Rust implementations return a `Codec` instance from the
/// `spec` function of `CodecTrait`. Plugins provide a JSON or YAML serialization
/// as part of their manifest.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(default, rename_all = "camelCase", crate = "common::serde")]
pub struct Codec {
    /// A list of format names (or aliases) that the codec can decode from
    /// or encode to
    pub formats: Vec<String>,

    /// Whether the codec supports decoding from string content
    pub from_string: bool,

    /// Whether the codec supports decoding from a file system path
    pub from_path: bool,

    /// Whether the codec supports encoding to string content
    pub to_string: bool,

    /// Whether the codec supports encoding to a file system path
    pub to_path: bool,

    /// A list of root node types that the codec can encode / decode
    ///
    /// Most codecs usually only handle one root type e.g. `Article`.
    /// Used to provide a list of formats to the user that support
    /// the current document type.
    pub root_types: Vec<String>,

    /// A list of node types that the codec does not support
    ///
    /// Used to provide warnings to the user on potential loss of content
    /// when encoding using the codec.
    pub unsupported_types: Vec<String>,

    /// A list of node properties that the codec does not support
    ///
    /// The format for these strings is `<type>.<property>` e.g. `Article.funders`
    /// Used to provide warnings to the user on potential loss of content
    /// when encoding using the codec.
    pub unsupported_properties: Vec<String>,

    /// The status of the codec e.g. `alpha`, `beta`, `stable`
    pub status: String,
}

impl Default for Codec {
    fn default() -> Self {
        Self {
            formats: vec![],
            from_string: true,
            from_path: true,
            to_string: true,
            to_path: true,
            root_types: vec_string!["Article"],
            unsupported_types: vec![],
            unsupported_properties: vec![],
            status: "alpha".to_string(),
        }
    }
}

/// A trait for codecs
///
/// This trait can be used by Rust implementations of codecs, allowing them to
/// be compiled into the Stencila binaries.
///
/// It defines similar functions to `serde_json` (and other `serde_` crates) for
/// converting nodes to/from strings, files, readers etc.
#[async_trait]
pub trait CodecTrait {
    /// Get the [`Codec`] specification for this implementation
    fn spec() -> Codec;

    /// Decode a document node from a string
    fn from_str(_str: &str, _options: Option<DecodeOptions>) -> Result<Node> {
        bail!("Decoding from string is not implemented for this format")
    }

    /// Decode a document node from a string asynchronously
    async fn from_str_async(str: &str, options: Option<DecodeOptions>) -> Result<Node> {
        Self::from_str(str, options)
    }

    /// Decode a document node from a `BufReader`
    async fn from_buffer<T: AsyncRead>(
        reader: &mut BufReader<T>,
        options: Option<DecodeOptions>,
    ) -> Result<Node>
    where
        T: Send + Sync + Unpin,
    {
        let mut content = String::new();
        reader.read_to_string(&mut content).await?;
        Self::from_str_async(&content, options).await
    }

    /// Decode a document node from a file
    ///
    /// This function reads the file as a string and passes that on to `from_str`
    /// for decoding. If working with binary formats, you should override this function
    /// to read the file as bytes instead.
    async fn from_file(file: &mut File, options: Option<DecodeOptions>) -> Result<Node> {
        let mut content = String::new();
        file.read_to_string(&mut content).await?;
        Self::from_str_async(&content, options).await
    }

    /// Decode a document node from a file system path
    async fn from_path(path: &Path, options: Option<DecodeOptions>) -> Result<Node> {
        let mut file = File::open(path).await?;
        Self::from_file(&mut file, options).await
    }

    /// Encode a document node to a string
    fn to_string(_node: &Node, _options: Option<EncodeOptions>) -> Result<String> {
        bail!("Encoding to string is not implemented for this format")
    }

    /// Encode a document node to a string asynchronously
    async fn to_string_async(node: &Node, options: Option<EncodeOptions>) -> Result<String> {
        Self::to_string(node, options)
    }

    /// Encode a document node to a `BufWriter`
    async fn to_buffer<T: AsyncWrite>(
        node: &Node,
        writer: &mut BufWriter<T>,
        options: Option<EncodeOptions>,
    ) -> Result<()>
    where
        T: Send + Sync + Unpin,
    {
        let content = Self::to_string_async(node, options).await?;
        writer.write_all(content.as_bytes()).await?;
        Ok(())
    }

    /// Encode a document node to a file
    async fn to_file(node: &Node, file: &mut File, options: Option<EncodeOptions>) -> Result<()> {
        let content = Self::to_string_async(node, options).await?;
        file.write_all(content.as_bytes()).await?;
        Ok(())
    }

    /// Encode a document node to a file system path
    async fn to_path(node: &Node, path: &Path, options: Option<EncodeOptions>) -> Result<()> {
        if let Some(parent) = path.parent() {
            create_dir_all(parent)?
        }
        let mut file = File::create(path).await?;
        Self::to_file(node, &mut file, options).await
    }
}

/// Decoding options
///
/// Decoding functions (including those in plugins) are encouraged to respect these options
/// but are not required to.
#[derive(Clone, Default)]
pub struct DecodeOptions {
    /// The format of the input to be decoded
    ///
    /// Most codecs only decode one format. However, for those that handle multiple
    /// format it may be necessary to specify this option.
    pub format: Option<String>,
}

/// Encoding options
///
/// Encoding functions (including those in plugins) are encouraged to respect these options
/// but are not required to. Indeed, some options do not apply for some formats.
/// For example, a PDF is always `standalone` (so if that option is set to `false`, it will be ignored).
/// Futhermore, some combinations of options are ineffectual e.g. a `theme` when `standalone: false`
#[derive(Clone)]
pub struct EncodeOptions {
    /// Whether to encode in compact form
    ///
    /// Some formats (e.g HTML and JSON) can be encoded in either compact
    /// or "pretty-printed" (e.g. indented) forms.
    pub compact: bool,

    /// Whether to ensure that the encoded document is standalone
    ///
    /// Some formats (e.g. Markdown, DOCX) are always standalone.
    /// Others can be fragments, or standalone documents (e.g HTML).
    pub standalone: bool,

    /// Whether to bundle local media files into the encoded document
    ///
    /// Some formats (e.g. DOCX, PDF) always bundle. For HTML, bundling means
    /// including media as data URIs rather than links to files.
    pub bundle: bool,

    /// Whether to use Stencila Cloud as a remote storage for document nodes
    /// that can not be wholly represented within the format
    ///
    /// Some formats (e.g. Google Docs) are not able to fully represent some types
    /// of nodes (e.g. CodeChunks). In these cases we store the node on Stencila Cloud
    /// with a unique id and key and insert a link to it from within the format.
    pub rpng_content: bool,
    pub rpng_link: bool,

    /// The theme to apply to the encoded document
    ///
    /// Only applies to standalone HTML and formats derived from it (e.g. PDF, PNG).
    pub theme: Option<String>,

    /// Whether Javascript for Web Components should be loaded or not
    ///
    /// Only applies to standalone HTML and formats derived from it (e.g. PDF, PNG).
    pub components: bool,

    /// The format to encode to
    ///
    /// Most codecs only encode to one format. However, for those that handle multiple
    /// formats it may be necessary to specify this option.
    pub format: Option<String>,
}

impl Default for EncodeOptions {
    fn default() -> Self {
        Self {
            compact: true,
            standalone: false,
            bundle: false,
            rpng_content: false,
            rpng_link: false,
            theme: None,
            components: true,
            format: None,
        }
    }
}
