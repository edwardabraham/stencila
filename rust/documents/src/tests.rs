use std::{path::PathBuf, sync::Arc};

use codec::CodecTrait;
use codec_md::MdCodec;
use common::{
    eyre::Result,
    serde_json,
    tokio::{
        self,
        sync::{mpsc, RwLock},
    },
};
use graph::{Plan, PlanOptions, PlanOrdering};
use graph_triples::TagMap;
use kernels::{Kernel, KernelSpace, KernelType};
use node_address::Slot;
use node_patch::{Operation, Patch};
use stencila_schema::Node;
use test_snaps::{
    fixtures,
    insta::{self, assert_json_snapshot},
    snapshot_set_suffix,
};

use crate::{
    assemble::assemble,
    compile::compile,
    document::CallDocuments,
    execute::execute,
    messages::{CancelRequest, PatchRequest},
};

/// Higher level tests of the top level functions in this crate
#[tokio::test]
async fn md_articles() -> Result<()> {
    let fixtures = fixtures();

    // So that test results are not dependant upon the the machine the test is run on or how
    // the test is compiled only use built-in kernels
    let kernels: Vec<Kernel> = kernels::available()
        .await
        .into_iter()
        .filter(|kernel| matches!(kernel.r#type, KernelType::Builtin))
        .collect();

    let mut settings = insta::Settings::clone_current();
    settings.set_prepend_module_to_snapshot(false);

    #[allow(deprecated)] // Using bind_to_scope leads to renaming of snapshot files
    settings.bind_to_thread();

    for name in ["code.md", "code-relations.md"] {
        let path = fixtures.join("articles").join(name);

        // Load the article
        let root = Arc::new(RwLock::new(MdCodec::from_path(&path, None).await?));

        // Strip the fixtures path so it does not differ between machines
        let path = path.strip_prefix(&fixtures)?;
        let project = path.parent().unwrap();

        let (patch_request_sender, mut patch_request_receiver) =
            mpsc::unbounded_channel::<PatchRequest>();
        tokio::spawn(async move {
            while let Some(_request) = patch_request_receiver.recv().await {
                // Ignore for this test
            }
        });

        let call_docs = Arc::new(RwLock::new(CallDocuments::default()));
        let tag_map = Arc::new(RwLock::new(TagMap::default()));

        // Assemble the article and snapshot the result
        let address_map = assemble(path, &root, &call_docs, &patch_request_sender).await?;
        snapshot_set_suffix(&[name, "-assemble"].concat(), || {
            assert_json_snapshot!(&address_map)
        });
        let address_map = Arc::new(RwLock::new(address_map));

        // Compile the article and snapshot the result
        let graph = compile(
            path,
            project,
            &root,
            &address_map,
            &tag_map,
            &patch_request_sender,
        )
        .await?;
        snapshot_set_suffix(&[name, "-compile"].concat(), || {
            assert_json_snapshot!(&graph)
        });

        // Generate various execution plans for the article using alternative options
        // and snapshot them all. Always specify `max_concurrency` to avoid differences
        // due to machine (number of CPUs)
        for (suffix, options) in [
            (
                "appearance",
                PlanOptions {
                    ordering: PlanOrdering::Appearance,
                    max_concurrency: 10,
                },
            ),
            (
                "appearance-concurrency-0",
                PlanOptions {
                    ordering: PlanOrdering::Appearance,
                    max_concurrency: 0,
                },
            ),
            (
                "topological",
                PlanOptions {
                    ordering: PlanOrdering::Topological,
                    max_concurrency: 10,
                },
            ),
            (
                "topological-concurrency-0",
                PlanOptions {
                    ordering: PlanOrdering::Topological,
                    max_concurrency: 0,
                },
            ),
        ] {
            let plan = graph
                .plan(None, Some(kernels.clone()), None, Some(options))
                .await?;
            snapshot_set_suffix(&[name, "-", suffix].concat(), || {
                assert_json_snapshot!(&plan)
            });
        }

        // Execute the article (with topological execution plan) and snapshot the resultant patches
        let plan = graph
            .plan(
                None,
                Some(kernels.clone()),
                None,
                Some(PlanOptions {
                    ordering: PlanOrdering::Topological,
                    max_concurrency: 10,
                }),
            )
            .await?;

        let (patch_request_sender, mut patch_request_receiver) =
            mpsc::unbounded_channel::<PatchRequest>();
        let patches = tokio::spawn(async move {
            let mut patches = Vec::new();
            while let Some(PatchRequest { mut patch, .. }) = patch_request_receiver.recv().await {
                // Redact execute time and duration from patch because they will
                // change across test runs
                for op in patch.ops.iter_mut() {
                    if let Operation::Add { address, value, .. } = op {
                        if let Some(Slot::Name(name)) = address.back() {
                            if name == "execute_ended" || name == "execute_duration" {
                                *value = Box::new("<redacted>".to_string());
                            }
                        }
                    }
                }
                patches.push(patch);
            }
            patches
        });

        let (cancel_request_sender, mut cancel_request_receiver) =
            mpsc::channel::<CancelRequest>(1);

        execute(
            &plan,
            &root,
            &address_map,
            &tag_map,
            &Arc::new(RwLock::new(KernelSpace::new(None, None))),
            &call_docs,
            &patch_request_sender,
            &mut cancel_request_receiver,
        )
        .await?;

        drop(patch_request_sender);
        drop(cancel_request_sender);

        let _patches = patches.await?;
        /*
        Snapshotting of patches turned off for now because order is not-deterministic

        snapshot_set_suffix(&[name, "-patches"].concat(), || {
            assert_json_snapshot!(&patches);
        });
        */
    }

    Ok(())
}

/// Regression test for executing a `CreativeWork`
///
/// Normally we execute a type of creative work such as an `Article`.
/// But sometimes, when you have a flat array of `Node`s, it is necessary
/// to use a `CreativeWork`. The first time we tried to do that it failed.
/// This just tests that there is a patch with an "execute_ended" op
#[tokio::test]
async fn regression_creative_work() -> Result<()> {
    let node = serde_json::from_value(serde_json::json!({
        "type": "CreativeWork",
        "content": [
            {
                "type": "CodeChunk",
                "text": "2 * 2",
                "programmingLanguage": "calc"
            }
        ]
    }))?;

    let (_plan, patches) = assemble_compile_plan_execute(node).await?;
    let was_executed = patches
        .iter()
        .flat_map(|patch| &patch.ops)
        .any(|op| match op {
            Operation::Add { address, .. } => match &address[0] {
                Slot::Name(name) => name == "execute_ended",
                _ => false,
            },
            _ => false,
        });
    assert!(was_executed);

    Ok(())
}

/// Convenience function to compile, plan and execute a node
///
/// Returns the plan and generated patches.
async fn assemble_compile_plan_execute(node: Node) -> Result<(Plan, Vec<Patch>)> {
    let root = Arc::new(RwLock::new(node));
    let call_docs = Arc::new(RwLock::new(CallDocuments::default()));
    let tags = Arc::new(RwLock::new(TagMap::default()));

    let (patch_request_sender, mut patch_request_receiver) =
        mpsc::unbounded_channel::<PatchRequest>();
    let patches = tokio::spawn(async move {
        let mut patches = Vec::new();
        while let Some(request) = patch_request_receiver.recv().await {
            patches.push(request.patch)
        }
        patches
    });

    let (_cancel_request_sender, mut cancel_request_receiver) = mpsc::channel::<CancelRequest>(1);

    let address_map = assemble(&PathBuf::new(), &root, &call_docs, &patch_request_sender).await?;
    let address_map = &Arc::new(RwLock::new(address_map));

    let graph = compile(
        &PathBuf::new(),
        &PathBuf::new(),
        &root,
        address_map,
        &tags,
        &patch_request_sender,
    )
    .await?;

    let plan = graph.plan(None, None, None, None).await?;

    execute(
        &plan,
        &root,
        address_map,
        &tags,
        &Arc::new(RwLock::new(KernelSpace::new(None, None))),
        &call_docs,
        &patch_request_sender,
        &mut cancel_request_receiver,
    )
    .await?;

    drop(patch_request_sender);
    let patches = patches.await?;

    Ok((plan, patches))
}
