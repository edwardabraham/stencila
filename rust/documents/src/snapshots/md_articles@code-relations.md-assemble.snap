---
source: rust/documents/src/tests.rs
assertion_line: 74
expression: "&address_map"
---
{
  "cf-1": [
    "content",
    0,
    "content",
    1
  ],
  "cf-2": [
    "content",
    0,
    "content",
    3
  ],
  "cf-3": [
    "content",
    0,
    "content",
    5
  ],
  "cf-4": [
    "content",
    0,
    "content",
    7
  ],
  "cf-5": [
    "content",
    1,
    "content",
    1
  ],
  "cf-6": [
    "content",
    1,
    "content",
    3
  ],
  "cf-7": [
    "content",
    1,
    "content",
    5
  ],
  "cc-1": [
    "content",
    2
  ],
  "cf-8": [
    "content",
    3,
    "content",
    1
  ],
  "cc-2": [
    "content",
    4
  ],
  "cf-9": [
    "content",
    5,
    "content",
    1
  ],
  "ce-1": [
    "content",
    5,
    "content",
    3
  ],
  "ce-2": [
    "content",
    5,
    "content",
    5
  ],
  "cf-10": [
    "content",
    6,
    "content",
    1
  ],
  "cf-11": [
    "content",
    6,
    "content",
    3
  ],
  "cc-3": [
    "content",
    7
  ],
  "ce-3": [
    "content",
    8,
    "content",
    1
  ],
  "ce-4": [
    "content",
    8,
    "content",
    3
  ],
  "cf-12": [
    "content",
    9,
    "content",
    1
  ],
  "cf-13": [
    "content",
    9,
    "content",
    3
  ],
  "pa-1": [
    "content",
    9,
    "content",
    5
  ],
  "cc-4": [
    "content",
    11
  ]
}
